Engine/TransactionHandlers/CancelPendingOrders.cs

```diff
@@ -0,0 +1,88 @@
+    /// Class used to keep track of CancelPending orders and their original or updated status
+    /// </summary>
+    public class CancelPendingOrders
+    {
+        private readonly ConcurrentDictionary<int, CancelPendingOrder> _cancelPendingOrders = new ConcurrentDictionary<int, CancelPendingOrder>();
+
+        /// <summary>
+        /// Amount of CancelPending Orders
+        /// </summary>
+        public int GetCancelPendingOrdersSize => _cancelPendingOrders.Count;
+
+        /// <summary>
+        /// Adds an order which will be canceled and we want to keep track of it Status in case of fallback
+        /// </summary>
+        /// <param name="orderId">The order id</param>
+        /// <param name="status">The order Status, before the cancel request</param>
+        public void Add(int orderId, OrderStatus status)
+        {
+            _cancelPendingOrders[orderId] = new CancelPendingOrder { Status = status };
+        }
+
+        /// <summary>
+        /// Updates an order that is pending to be canceled.
+        /// </summary>
+        /// <param name="newStatus">The new status of the order. If its OrderStatus.Canceled or OrderStatus.Filled it will be removed</param>
+        /// <param name="orderId">The id of the order</param>
+        public void UpdateOrRemove(OrderStatus newStatus, int orderId)
+        {
+            CancelPendingOrder cancelPendingOrder;
+            if (_cancelPendingOrders.TryGetValue(orderId, out cancelPendingOrder))
+            {
+                lock (cancelPendingOrder)
+                {
+                    if (newStatus == OrderStatus.Canceled || newStatus == OrderStatus.Filled)
+                    {
+                        RemoveOrderFromCollection(orderId);
+                    }
+                    else if (newStatus != OrderStatus.CancelPending)
+                    {
+                        // check again because it might of been removed by the failed to cancel event
+                        if (_cancelPendingOrders.ContainsKey(orderId))
+                        {
+                            _cancelPendingOrders[orderId].Status = newStatus;
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Removes an order which we failed to cancel and falls back the order Status to previous value
+        /// </summary>
+        /// <param name="order">The order that failed to be canceled</param>
+        public void RemoveAndFallback(Order order)
+        {
+            CancelPendingOrder cancelPendingOrder;
+            if (_cancelPendingOrders.TryGetValue(order.Id, out cancelPendingOrder))
+            {
+                lock (cancelPendingOrder)
+                {
+                    if (_cancelPendingOrders.ContainsKey(order.Id))
+                    {
+                        // update Status before removing from _cancelPendingOrders
+                        order.Status = _cancelPendingOrders[order.Id].Status;
```

Comment:
This is tempting, I also thought about it, but it will bring some issues:
A- From the user perspective they will be receiving again an old event (since it will fallback to the original or updated status), and it will mean adding logic to `HandleOrderEvent()` to avoid sending wrong information (or doing things twice).
B- I believe we also could have a threading issue. If we call the `HandleOrderEvent()` after the cancellation failure and there is an incoming call from the broker with `Filled` status, we will have two thread working on the same status.